hail/python/hailtop/aiotools/fs.py

```diff
@@ -737,22 +814,15 @@ async def remove(self, url: str) -> None:
         fs = self._get_fs(url)
         return await fs.remove(url)
 
-    async def rmtree(self, url: str) -> None:
+    async def rmtree(self, sema: asyncio.Semaphore, url: str) -> None:
         fs = self._get_fs(url)
-        return await fs.rmtree(url)
+        return await fs.rmtree(sema, url)
 
     async def close(self) -> None:
         for fs in self._filesystems:
             await fs.close()
 
-    async def copy(self, transfer: Union[Transfer, List[Transfer]], raise_first_exception=True):
-        worker_pool = AsyncWorkerPool(50)
-        try:
-            copier = Copier(self)
-            copy_report = CopyReport(transfer)
-            await copier.copy(worker_pool, copy_report, transfer)
-            if raise_first_exception:
-                copy_report.raise_first_exception()
-            return copy_report
-        finally:
-            worker_pool.shutdown()
+    async def copy(self, sema: asyncio.Semaphore, transfer: Union[Transfer, List[Transfer]], return_exceptions=False):
```

Comment:
Should we make the semaphore optional here and make the default 50? Same question about rmtree above. Then you wouldn't need to change all the test interfaces.

Reply:
So I thought about this, but I'd like to hold off until I understand the right design better.  I think defaulting to create a new semaphore is dangerous, because if you leave it off, you accidentally increase the amount of parallelism being used.  I'd rather have the right/safe interface for the normal code and let the tests be a bit more complicated.